### ISSUE TITLE
Global usage of static GQL in TypeMapper

### ISSUE THREAD

**Issue by yoadsn** [graphql-compose/graphql-compose#108]
@nodkz I have given more thought to the latest additions to TypeMapper functionality following #107.

For others who read - in these changes - the usage of GQL to lookup existing types is useful for us when configuring output/input fields and args trying to reuse types added to GQL with `GQL.add()` previously - by specifying the name as a string.

My concern with this addition is the dependency now TypeMapper has (and transitively TypeComposer) on the global static instance of GQC.

The result is that we cannot build two separate schemas within the same node process.
A use case for that could be exposing two different schemas from the same node server for different usages (server-to-server vs server-to-client) - If those contain the same type names, the type would be reused which may not be the correct result. Perhaps there would be other problems.

But, what do you think?
Could it be better if TypeMapper can get an instance of `TypeStorage` that when provided would be used instead of the global GQC static `TypeStorage`?

P.S
It seems to me GQC and the TypeStorage it extends need not be static at all in the long run and should be created (maybe explicitly if not provided) for the composition of an entire schema.
I understand this complicates the usage of `graphql-compose` but maybe the API can be simpler when only creating a single schema.

Same goes for TypeMapper which internally uses a static TypeMapper instance to store parsed AST's and definitions. (was also added for the same purpose?)

We can have a conversation about what could be the way to move away from a static TypeStorage like GQC in the future. This would make the library more robust.

**Comment by nodkz** [comment]
Majority users will not combine several schemas under one node process. Or will use different names for types with different behavior. Currently, it's not a problem using one type in different schemas with the same behavior.  Anyway current new behavior will be very useful for majority. Tradeoffs 👐 

Problems will have a minority of users, which want to use same type names with different implementations. A very bugus way from my point of view. But why not if somebody decides to do it.   

I have a solution/proposal for such schemas construction with minimal changes in API. It can be done via inherited TCs, TypeMapper and other classes dependent on one type storage. 

Standart using of graphql-compose:
```js
// ArticleTC.js
import { TypeComposer, Resolver, ... } from 'graphql-compose';
const ArticleTC = TypeComposer.create(...);
```

PROPOSAL for obtaining inherited classes with own storage:
```js 
// gqc1.js
import { createNamespace } from 'graphql-compose';

// **createNamespace** - is not a good name. Need to find a better name.
const { GQC, TypeComposer, Resolver, ... } = createNamespace();

export {
  TypeComposer, Resolver, ... 
};
```
And ArticleTC maybe easy rewritten in following manner:
```diff
- import { TypeComposer, Resolver, ... } from 'graphql-compose';
+ import { TypeComposer, Resolver, ... } from './gqc1.js';
const ArticleTC = TypeComposer.create(...);
```

So all new inherited classes will have statically defined own global type storage and can be used without any API changes. Just need to change imports.

Thoughts?

----
PS. It might sound weird at first, but as a fast solution we can publish several packages to npm with following names `graphql-compose-1`, `graphql-compose-2`, `graphql-compose-3`, `graphql-compose-4`. 🤣 But it will really work 😈

**Comment by yoadsn** [comment]
> we can publish several packages to npm...

And you say the I'm crazy... 😆 

I realize this might not be a common use case, but the fact is that when I see a global state I die a little inside. I would say better be explicit about the state then implicit. For example, the "default connection" of mongoose is good for beginners - but horrible in terms of design IMO and useless for advanced scenarios.

Your proposal sounds good to me, every type would wrap around the storage when created via what you called "namespace". In fact, maybe a good name could be "CompositionContainer".

I cannot think of a better way to do it myself with what JS has to offer. (until there is a concept of "thread context" in the runtime). And also without breaking the existing API.

We can experiment on a branch - but if you say you do not think it's that important we are both busy so we can leave that until a real use-case is up.

I guess when "schema stitching" abilities make their way into 'graphql-compose' family, this may be more important - having to generate multiple schemas first and than merge them. They need to live in different "Containers" before the merge process resolves duplicate names (with namespacing?) and conflicting type implementations.

**Comment by nodkz** [comment]
> For example, the "default connection" of mongoose is good for beginners - but horrible in terms of design IMO and useless for advanced scenarios.

Good catch. Mongoose solves this via a `connection`. You may take schema and connection and create a models with any advanced scenarios. Eg:
```js
const connection1 = mongoose.createConnection();
const connection2 = mongoose.createConnection();
const UserSchema = new mongoose.Schema({ ... });
export const UserDB1 = connection1.model('User', UserSchema);
export const UserDB2 = connection2.model('User', UserSchema);
// establish real connection to database somewhere later 
connection1.openUri(uri, opts);
connection2.openUri(uri, opts);
```
For newcomers, it's not obvious, but it's problem of documentation. In past I also thought that mongoose has not good design, but with time I understood all deepness of thoughts. And now I like it.

-----

I don't like nor "namespace", nor "CompositionContainer". And thank you a lot, your words bring me, as I think, a good idea and naming. And it's really trivial and obvious.  Our root of a problem around multiple Schemas? Right?

### So let it have `SchemaComposer` name. 
We already have TypeCompose, InputTypeComposer, EnumTypeComposer. Why do we not have SchemaComposer yet? 🤯

```js
import { SchemaComposer } from 'graphql-compose';

const SC = new SchemaComposer();
const { TypeComposer, Resolver, InputTypeComposer, TypeStorage, TypeMapper, EnumTypeComposer } = SC;
// ... type generation with unchanged API

const SC2 = new SchemaComposer(); // with own type storage
const SC3 = new SchemaComposer(); // with another type storage
```

So current exports from `graphql-compose` will return "default SchemaComposer". But if you need to create another schema, just import `SchemaComposer`, and create a new schema. Almost like it done in mongoose 😈

**Comment by yoadsn** [comment]
Awesome - totally agree! 💯 

I can give it a shot (not urgent I guess.. but good to start thinking with code) - although I keep having the problem of contributing to this projects without having a good understanding of flow types. We do not use that internally. Still, worth trying.

**Comment by nodkz** [comment]
And `GQC` must die ;)
SchemaComposer takes his place.

**Comment by nodkz** [comment]
Flowtype is awesome. It really pains to start with them, but after several months it starts to save a lot of time - type checking, autocompletion. After migration to RelayModern we also get type checking between graphql requests and components - catch and fix tons of runtime errors.

I don't know how I lived before without Flow.

**Comment by yoadsn** [comment]
Good place to start is contributing to GQC!

**Comment by nodkz** [comment]
I'll try to record a video about flow in action, if i have free time tomorrow.

Video is worth a thousand words.

**Comment by nodkz** [comment]
@yoadsn sorry no video today.

But I have something better ;)
Take a look at #109

**Comment by yoadsn** [comment]
@nodkz OMG.

**Comment by yoadsn** [comment]
Addressed in #109
